src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -588,19 +607,138 @@ public async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Compilation co
 
-        public ImmutableArray<Diagnostic> DequeueLocalDiagnostics(DiagnosticAnalyzer analyzer, bool syntax, Compilation compilation)
+        public void ApplyProgrammaticSuppressions(DiagnosticBag reportedDiagnostics, Compilation compilation)
+        {
+            Debug.Assert(!reportedDiagnostics.IsEmptyWithoutResolution);
+            if (!_hasDiagnosticSuppressors)
+            {
+                return;
+            }
+
+            var newDiagnostics = ApplyProgrammaticSuppressionsCore(reportedDiagnostics.ToReadOnly(), compilation);
+            reportedDiagnostics.Clear();
+            reportedDiagnostics.AddRange(newDiagnostics);
+        }
+
+        public ImmutableArray<Diagnostic> ApplyProgrammaticSuppressions(ImmutableArray<Diagnostic> reportedDiagnostics, Compilation compilation)
+        {
+            if (reportedDiagnostics.IsEmpty ||
+                !_hasDiagnosticSuppressors)
+            {
+                return reportedDiagnostics;
+            }
+
+            return ApplyProgrammaticSuppressionsCore(reportedDiagnostics, compilation);
+        }
+
+        private ImmutableArray<Diagnostic> ApplyProgrammaticSuppressionsCore(ImmutableArray<Diagnostic> reportedDiagnostics, Compilation compilation)
+        {
+            Debug.Assert(_hasDiagnosticSuppressors);
+            Debug.Assert(!reportedDiagnostics.IsEmpty);
+            Debug.Assert(_programmaticSuppressions != null);
+
+            // We do not allow analyzer based suppressions for following category of diagnostics:
+            //  1. Diagnostics which are already suppressed in source via pragma/suppress message attribute.
+            //  2. Diagnostics explicitly tagged as not configurable by analyzer authors - this includes compiler error diagnostics.
+            //  3. Diagnostics which are marked as error by default by diagnostic authors.
+            var suppressableDiagnostics = reportedDiagnostics.WhereAsArray(d => !d.IsSuppressed &&
+                                                                                !d.IsNotConfigurable() &&
+                                                                                d.DefaultSeverity != DiagnosticSeverity.Error);
+            if (suppressableDiagnostics.IsEmpty)
+            {
+                return reportedDiagnostics;
+            }
+
+            ExecuteSuppressionActions(suppressableDiagnostics, concurrent: compilation.Options.ConcurrentBuild);
+            if (_programmaticSuppressions.IsEmpty)
+            {
+                return reportedDiagnostics;
+            }
+
+            var builder = ArrayBuilder<Diagnostic>.GetInstance(reportedDiagnostics.Length);
+
+            var programmaticSuppressionsByDiagnostic = _programmaticSuppressions.GroupBy(s => s.SuppressedDiagnostic)
+                .ToImmutableDictionary(g => g.Key,
+                                       g => new ProgrammaticSuppressionInfo(g.Select(s => (s.Descriptor.Id, s.Descriptor.Justification)).ToImmutableSortedSet()));
```

Comment:
>ToImmutableSortedSet [](start = 137, length = 20)

What is the sort order here and why is it important?

Reply:
You are correct. Initially we had proposed to create/log a single suppression diagnostic for every programmatically suppressed diagnostic, whose message would list all the suppression IDs and justifications, which would need to be sorted. However, the final iteration of the design we decided to report one suppression diagnostic per suppression, so the sorting is not required anymore. I will switch this to a regular `ImmutableHashSet` and also fix the tests for multiple suppression scenario to do the sorting for determinism.